### ISSUE TITLE
How to ensure that occlum code is trusted?

### ISSUE THREAD

**Issue by SHiftLin** [occlum/occlum#714]
Hi, I read occlum's paper and tried occlum, but I have a question about it.
Suppose a user wants to run a trusted program to process his private data on a cloud machine. But the program does not use SGX and he does not trust the cloud machine and the OS. Thus, he uses occlum to run the program. But how could he ensure that the occlum installed on the cloud machine is correctly implemented ? 

I think occlum should provide its measurement (MRENCLAVE) by an out-of-band channel,  and the user should use remote attestation to attest the occlum and the program running inside the occlum? The user uploads his private data if the attestation succeed.

**Comment by guzongmin** [comment]
@SHiftLin When you are using Occlum on Could, you should provide the signed application + Occlum image. Which means you should sign them by yourself.

**Comment by SHiftLin** [comment]
Yes, but my application is running inside Occlum. Although I can trust my application and image, but how can I trust Occlum? How can I trust that Occlum actually run my application in an enclave or the Occlum itself is correctly implement?

**Comment by guzongmin** [comment]
@SHiftLin When you are signing the application with Occlum, the application hash becomes the part of Occlum MREnclave. You are able to check the MREnclave to confirm that.

**Comment by SHiftLin** [comment]
I see. Assuming a user trust SGX and Occlum, he generates the occlum instance at local, and runs the instance on the cloud.  So he can compare the mrenclave of enclave running on the cloud with one he generates at local.

**Comment by agowdamsft** [comment]
@SHiftLin to break it down. Trusted Code/Enclave App/Container for SGX = Your business logic application code + SGX Code Wrapper i.e Occlum. Occlum is fully Open sourced you have the ability to verify the code before trusting it as part of your software supply chain. Its similar to you trusting a NuGet package/PyPy packages as part of business logic code. Like @guzongmin mentioned, you have the ability to sign this full bundle and verify if this is exactly what is is running before you exchange any sensitive data. This process will put cloud operator or even you admins out of the trust boundary.

**Comment by lijun0314** [comment]
Hi, @guzongmin, 
Does the value of Occlum MREnclave include the hash of Occlum image? Or is it just the hash of enclave?

**Comment by zibinpan** [comment]
@SHiftLin @agowdamsft @guzongmin Sorry to bother you. Is there a way to solve my problem? Thank you.

**Comment by guzongmin** [comment]
Technically, you can get all those component codes used inside Occlum, including Occlum itself. After you review all the codes, you can build it and then save the MREnclave in the build process. When doing remote attestation, you can verify the MREnalve; if it is exactly the same as the one you saved, then the enclave is what you want.  The code you have reviewed is the first step, so it is as secure as you want.
